### PR TITLE
Add an MQTT broker.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/broker.toit [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+
+    if clean_session: broker.remove_session_ client_id
+
+    if reader_task_:
+      reader_task_.cancel
+      reader_task_ = null
+    if writer_task_:
+      writer_task_.cancel
+      writer_task_ = null
+
```
Comment (kasperl) [inline src/broker.toit:272]: Drop one newline.

File: src/broker.toit [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+  dispatch_incoming_publish packet/PublishPacket:
+    // There doesn't seem to be a rule which qos we should use if multiple
+    // subscriptions match. We thus use the one from the most specialized.
+    subscription_tree_.do --most_specialized packet.topic: | subscription_max_qos |
+      qos := min packet.qos subscription_max_qos
+      NO_PACKET_ID ::= -1  // See $PublishPacket.with.
+      packet_id := qos > 0 ? next_packet_id_++ : NO_PACKET_ID
+      send_ (packet.with --packet_id=packet_id --qos=qos)
+
```
Comment (kasperl) [inline src/broker.toit:288]: Drop one newline.

File: src/broker.toit [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+/** An unbounded channel for publish messages. */
+monitor PublishChannel_:
+  // Messages that haven't been sent yet.
+  queued_ /Deque := Deque
+
+  next -> Packet?:
```
Comment (kasperl) [inline src/broker.toit:295]:
```suggestion
  next -> Packet:
```

File: src/broker.toit [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    // There doesn't seem to be a rule which qos we should use if multiple
+    // subscriptions match. We thus use the one from the most specialized.
+    subscription_tree_.do --most_specialized packet.topic: | subscription_max_qos |
+      qos := min packet.qos subscription_max_qos
+      NO_PACKET_ID ::= -1  // See $PublishPacket.with.
+      packet_id := qos > 0 ? next_packet_id_++ : NO_PACKET_ID
+      send_ (packet.with --packet_id=packet_id --qos=qos)
+
+
+/** An unbounded channel for publish messages. */
```
Comment (kasperl) [inline src/broker.toit:290]: This could also be a semaphore and a deque.

File: src/broker.toit [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+          session.dispatch_incoming_publish packet
+
+    try:
+      server_transport_.listen::
+        logger_.info "connection established"
+        connection := Connection_ it
+        exception := catch --trace:
+          packet := connection.read
+          if not packet:
+            logger_.info "Connection was closed"
```
Comment (kasperl) [inline src/broker.toit:332]:
```suggestion
            logger_.info "connection was closed"
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/broker.toit (hunk 1 above)
Comment (floitsch) [reply src/broker.toit:273]: done.

File: src/broker.toit (hunk 2 above)
Comment (floitsch) [reply src/broker.toit]: done.

File: src/broker.toit (hunk 4 above)
Comment (floitsch) [reply src/broker.toit:289]: done.

**Pull request merged**